### ISSUE TITLE
Question: Is there a way to define circularly dependent type variables?

### ISSUE THREAD

**Issue by Matthewacon** [raphw/byte-buddy#485]
Keeping in mind that all of this code is valid, and will compile: given a starting interface:
```
public interface ITest {}
```
would it be possible to redefine `ITest` to the following:
```
public interface ITest<T extends Enum<T> & ITest<T>> {
 T exampleMethod();
}
```
The main issue that I'm encountering is that `DynamicType.Builder.typeVariable()`,  `TypeDescription.Generic.Builder.parameterizedType()` and `Builder.defineMethod()` do not accpet `TypeVariableDefinition`s as parameters. Is there a way to achive the desired outcome?

**Comment by raphw** [comment]
It is possible but you need to apply a little trick to avoid Byte Buddy's validation of the `ITest` during the creation of the parameterized type:

```java

DynamicType.Unloaded<ITest> type = new ByteBuddy()
.redefine(ITest.class)
.typeVariable("T",
  TypeDescription.Generic.Builder.parameterizedType(
    new TypeDescription.ForLoadedType(Enum.class),
    TypeDescription.Generic.Builder.typeVariable("T").build()).build(),
    TypeDescription.Generic.Builder.parameterizedType(
      new TypeDescription.ForLoadedType(ITest.class) {
        @Override
        public TypeList.Generic getTypeVariables() {
          return new TypeList.Generic.Explicit(TypeDescription.Generic.Builder.typeVariable("T").build());
        }
      },
      TypeDescription.Generic.Builder.typeVariable("T").build()).build()
  )
  .defineMethod("exampleMethod",
      TypeDescription.Generic.Builder.typeVariable("T").build(),
      Visibility.PUBLIC)
  .withoutCode()  
  .make();
```

**Comment by Matthewacon** [comment]
Thanks for the reply! I've got three more questions, one related to type parameters, one related to setting access flags and one about default members values in annotations. Since I couldn't find anything, I felt it appropriate to post them here.

First question: How would you define a method with type parameters. For example:
```
interface SomeInterface {
 <T extends SomeClass<T> & SomeInterface> myMethodStub();
}
```

Second Question: Does ByteBuddy support directly modifying access flags like in asm? Given an example class, such as the following:
```
public class ExampleClass {
 public static final ExampleClass SOME_CONSTANT;
}
```
how would I manually set the class access flags to `ACC_PUBLIC` and `ACC_ENUM`? Same for the field, `ACC_PUBLIC`, `ACC_STATIC`, `ACC_FINAL` and `ACC_ENUM`. To be clear, I do not want to simply define an enum, using `new ByteBuddy().makeEnumeration(String...)`, I want to redefine a pre-existing class and convert it into an enum. 

Third Question: How would you set a default value for an annotation? For example:
```
public @interface ExampleAnnotation {
 String value default "This is a default value";
}
```

**Comment by raphw** [comment]
As for method type variables, the DSL supports it. Typically, your IDE can even help you with code completion here. You would specify the type variable after defining the method via `defineMethod`.

You can transform modifiers via the `field` step. Via the `method` step, you can match preexisting methods or via `defineMethod` you could define one. Annotation properties are nothing but methods where there is a `defaultValue` step where you can define such a value.

**Comment by Matthewacon** [comment]
Thanks for all the help!

**Comment by Matthewacon** [comment]
In #402 you stated that a static initializer could be intercepted by `builder.executable(ElementMatchers.isTypeInitializer()).intercept(...)`. Could you clarify which builder you were referencing in this issue?

**Comment by raphw** [comment]
This is intentional and there is currently no way to suppress the original initializer. The reason for this is that a skipped execution of the initializer can break a class. You can however register an AsmVisitorWrapper to drop the original initializer by returning `null` from `visitMethod` once the method is requested.

**Comment by Matthewacon** [comment]
Is there, then, a way to control the order of invocation? For my purposes I require that the interceptor be invoked before the original initializer.

**Comment by raphw** [comment]
No, unfortunately.

**Comment by Matthewacon** [comment]
Apologies for the long gap in issue activity, I should've closed this issue a month ago. I found that, through the `Advice` API, I was able to perform static initializer interception in the order that I required. That being said, I would like to make the last topic of this issue a feature request; would it be difficult to implement, in the high-level API, fine control over static initializers? Features such as execution order, similar to the `Advice` API but for existing static initializers, and suppression or redefinition, such as in the interception manor. And on a related note, when you stated 

> a skipped execution of the initializer can break a class

would that not be true for any form of interception, where the target method is replaced? What amplifies the severity of intercepting a static initializer and not invoking the original code as opposed to doing the same to a static or instance bound method? If it's a matter of implementation difficulty, then no worries, just curious.
